src/sessions.ts

```diff
@@ -801,28 +820,35 @@ function applySession(
   // first apply non-transaction-specific sessions data
   const inTransaction = session.inTransaction() || isTransactionCommand(command);
   const isRetryableWrite = options?.willRetryWrite || false;
-  const shouldApplyReadConcern = commandSupportsReadConcern(command, options);
 
   if (serverSession.txnNumber && (isRetryableWrite || inTransaction)) {
     command.txnNumber = Long.fromNumber(serverSession.txnNumber);
   }
 
-  // now attempt to apply transaction-specific sessions data
   if (!inTransaction) {
     if (session.transaction.state !== TxnState.NO_TRANSACTION) {
       session.transaction.transition(TxnState.NO_TRANSACTION);
     }
 
-    // TODO: the following should only be applied to read operation per spec.
-    // for causal consistency
-    if (session.supports.causalConsistency && session.operationTime && shouldApplyReadConcern) {
+    if (
+      session.supports.causalConsistency &&
+      session.operationTime &&
+      commandSupportsReadConcern(command, options)
+    ) {
       command.readConcern = command.readConcern || {};
       Object.assign(command.readConcern, { afterClusterTime: session.operationTime });
+    } else if (session.snapshot) {
+      command.readConcern = command.readConcern || { level: ReadConcernLevel.snapshot }; // TODO: is there a better place to set this?
```

Comment:
Should we be using the ReadConcern proper class here? If it doesn't cause failures I think this style of usage is just legacy

Reply:
I think we should revisit this later because based on what's defined in the ReadConcern class proper, only the level is present, and based on the usage, there is ReadConcern defined for entire entities, like collections, and then there are specific read concerns passed with individual commands (containing things like the afterClusterTime and atClusterTime); from a typescript hinting perspective, it would not be helpful to mix the two since these additional properties can't be defined at a higher level... so we probably want to think more on how we want to design this internally going forward